update python release process

tmr232 opened the pull request:
I tried to make things a bit more modern and automated...

- Replaced old `setup.py` with `pyproject.toml` using uv and hatch
- Added a workflow for publishing to PyPI and set up trusted publishing for it
- Added a test workflow, so that we can manually publish to TestPyPI to verify things

I tested locally to see that the generated wheels look ok, but I don't have Ghidra installed, so I can't fully test it right now.

tmr232 committed:
Migrate from setup.py to UV and Hatch

tmr232 committed:
Add GH workflow for publishing the PyPI package

tmr232 commented:
@fmagin Can you take a look?

fmagin commented on .github/workflows/release-python.yml:
```diff
@@ -0,0 +1,36 @@
+on:
+  push:
+    tags:
+      - 'v*.*.*'
```
This would mean we try to publish the Python package for every tagged commit? Which would fail because we keep trying to re-publish the version 1.1.1 of the Python package? The Python package has changed roughly once in the last few years, while I keep updating the actual Ghidra plugin. I'm not sure if there is a good way to uncouple this. Maybe a different tag for Python releases?

fmagin reviewed (commented):
I think the current way that the python release is triggered would cause a failure the next time I try to update the Ghidra plugin without changing the Python package version. If those two need to be in sync and updated together then we should set the Python package version to the current Ghidra Plugin Version

tmr232 committed:
Infer Python version from package version

tmr232 commented:
> I think the current way that the python release is triggered would cause a failure the next time I try to update the Ghidra plugin without changing the Python package version. If those two need to be in sync and updated together then we should set the Python package version to the current Ghidra Plugin Version

I changed the `pyproject.toml` so that it uses the git tag for the Python package version. I think this should fix it?

fmagin commented:
Looks like that should fix. I did not know that this is possible, neat.

tmr232 commented:
Yeah, the new tooling is really great!

Can you build the wheels locally and see that they work for you? (`uv build` should do the trick)

tmr232 commented:
(And once that done, and we merge, we'll see how much of a mess I made with the GH actions :P

fmagin commented:
The built wheel works as far as I can tell. I guess with the new fancy tooling it might be sufficient to even just run `uv run ghidra-jupyter install-extension --ghidra ~/ghidras/ghidra_11.3_PUBLIC/` for testing this? but I also tested via the wheels, LGTM

tmr232 commented:
OK, I'm merging and we'll test the release...

tmr232 commented:
OK, something is _definitely_ wrong with the automatic publish, and I have no idea what and no energy to debug that now.
Should I just upload a new wheel, or can you do that?

fmagin commented:
Quick skim over the output: `ghidra_jupyter-0.0.post1.dev1+ga6b4a1d-py3-none-any.whl` looks like it's not finding the correct version number?

tmr232 commented:
This is probably why - 
```
/tmp/baipp-uv_cache_dir/builds-v0/.tmpRnDxFS/lib/python3.9/site-packages/setuptools_scm/git.py:167: UserWarning: "/home/runner/work/ghidra-jupyter-kotlin/ghidra-jupyter-kotlin" is shallow and may cause errors
  warnings.warn(f'"{wd.path}" is shallow and may cause errors')
```

I can fix that, but it doesn't explain why the upload failed.

fmagin commented:
Trying to manually publish fails with a very similar but more helpful error:
```
Uploading ghidra_jupyter-1.7.2.post1.dev3+gc356386.d20250809-py3-none-any.whl (27.6KiB)
error: Failed to publish `dist/ghidra_jupyter-1.7.2.post1.dev3+gc356386.d20250809-py3-none-any.whl` to https://upload.pypi.org/legacy/
  Caused by: Upload failed with status code 400 Bad Request. Server says: 400 The use of local versions in <Version('1.7.2.post1.dev3+gc356386.d20250809')> is not allowed. See https://packaging.python.org/specifications/core-metadata for more information.
```

tmr232 commented:
So we need to change the version scheme to something that matches - https://setuptools-scm.readthedocs.io/en/latest/extending/#setuptools_scmversion_scheme 

The "right tag" thing is now resolved (in the testing one, should migrate changes to the non-test one once we get it working fully)

fmagin commented:
`only-version` seems like the correct one?

tmr232 commented:
I say - use that, until it breaks :P

tmr232 commented:
@fmagin can you take care of it? I've had enough of this for today...

fmagin commented:
I did succeed in pushing a release manually at least: https://pypi.org/project/ghidra-jupyter/1.7.3/ but I'm not quite sure why this now worked

fmagin commented:
Okay, so as far as I understand what happened (and I'm not fully sure if this is correct):
The SCM version code uses the last tag as the base version, and if this is not the current commit then it appends `g$COMMIT_SHA`. if the working directory is dirty it also appends `d$DATE`. It seems like PyPI rejects versions with `g$COMMIT_ID` for some reason? So this only works if actually triggered via a tag on GitHub and not for a commit that isn't tagged.
